File: library/src/main/java/me/shkschneider/skeleton/ui/BottomSheet.java
package me.shkschneider.skeleton.ui;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Dialog;
import android.content.DialogInterface;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.annotation.UiThread;
import android.view.Gravity;
import android.view.View;
import android.view.Window;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import me.shkschneider.skeleton.R;
import me.shkschneider.skeleton.helper.ScreenHelper;

// <https://github.com/javiersantos/BottomDialogs>
public class BottomSheet {

    private Dialog mDialog;

    private BottomSheet() {
        // Forbidden
    }

    private BottomSheet(@NonNull final Builder builder) {
        mDialog = new Dialog(builder.activity, R.style.SkeletonTheme_BottomSheet);
        @SuppressLint("InflateParams")
        final View view = builder.activity.getLayoutInflater().inflate(R.layout.sk_bottomsheet, null);
        final ImageView icon = view.findViewById(R.id.sk_bottomsheet_icon);
        if (builder.icon != null) {
            icon.setVisibility(View.VISIBLE);
            icon.setImageDrawable(builder.icon);
        }
        final TextView title = view.findViewById(R.id.sk_bottomsheet_title);
        if (builder.title != null) {
            title.setText(builder.title);
        }
        final TextView content = view.findViewById(R.id.sk_bottomsheet_content);
        if (builder.content != null) {
            content.setText(builder.content);
        }
        final FrameLayout customView = view.findViewById(R.id.sk_bottomsheet_customView);
        if (builder.customView != null) {
            customView.addView(builder.customView);
            customView.setPadding(builder.paddingLeft, builder.paddingTop, builder.paddingRight, builder.paddingBottom);
        }
        final Button negative = view.findViewById(R.id.sk_bottomsheet_cancel);
        if (builder.negative != null) {
            negative.setVisibility(View.VISIBLE);
            negative.setText(builder.negative);
            negative.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (builder.negativeCallback != null)
                        builder.negativeCallback.onClick(mDialog, DialogInterface.BUTTON_NEGATIVE);
                    mDialog.dismiss();
                }
            });
        }
        final Button positive = view.findViewById(R.id.sk_bottomsheet_ok);
        if (builder.positive != null) {
            positive.setVisibility(View.VISIBLE);
            positive.setText(builder.positive);
            positive.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (builder.positiveCallback != null)
                        builder.positiveCallback.onClick(mDialog, DialogInterface.BUTTON_POSITIVE);
                    mDialog.dismiss();
                }
            });
        }
        mDialog.setContentView(view);
        mDialog.setCancelable(builder.cancelable);
        final Window window = mDialog.getWindow();
        if (window != null) {
            window.setLayout(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT);
            window.setGravity(Gravity.BOTTOM);
        }
    }

    @UiThread
    public void show() {
        mDialog.show();
    }

    @UiThread
    public void dismiss() {
        mDialog.dismiss();
    }

    public static class Builder {

        public final Activity activity;
        public Drawable icon;
        public String title;
        public String content;
        public String negative;
        public String positive;
        public DialogInterface.OnClickListener negativeCallback;
        public DialogInterface.OnClickListener positiveCallback;
        public View customView;
        public int paddingLeft;
        public int paddingTop;
        public int paddingRight;
        public int paddingBottom;
        public boolean cancelable;

        public Builder(@NonNull final Activity activity) {
            this.activity = activity;
            this.cancelable = true;
        }

        public Builder setTitle(@NonNull final String title) {
            this.title = title;
            return this;
        }

        public Builder setContent(@NonNull final String content) {
            this.content = content;
            return this;
        }

        public Builder setIcon(@NonNull final Drawable icon) {
            this.icon = icon;
            return this;
        }

        public Builder setPositive(@NonNull final String positive, final DialogInterface.OnClickListener onClickListener) {
            this.positive = positive;
            this.positiveCallback = onClickListener;
            return this;
        }

        public Builder setNegative(@NonNull final String negative, final DialogInterface.OnClickListener onClickListener) {
            this.negative = negative;
            this.negativeCallback = onClickListener;
            return this;
        }

        public Builder setCancelable(final boolean cancelable) {
            this.cancelable = cancelable;
            return this;
        }

        public Builder setCustomView(@NonNull final View customView) {
            this.customView = customView;
            this.paddingLeft = 0;
            this.paddingTop = 0;
            this.paddingRight = 0;
            this.paddingBottom = 0;
            return this;
        }

        public Builder setCustomView(@NonNull final View customView, final int left, final int top, final int right, final int bottom) {
            this.customView = customView;
            this.paddingLeft = ScreenHelper.pixelsFromDp(left);
            this.paddingTop = ScreenHelper.pixelsFromDp(top);
            this.paddingRight = ScreenHelper.pixelsFromDp(right);
            this.paddingBottom = ScreenHelper.pixelsFromDp(bottom);
            return this;
        }

        @UiThread
        public BottomSheet build() {
            return new BottomSheet(this);
        }

    }

}

File: library/src/main/java/me/shkschneider/skeleton/ui/BottomSheet.kt
package me.shkschneider.skeleton.ui

import android.app.Activity
import android.app.Dialog
import android.content.DialogInterface
import android.graphics.drawable.Drawable
import android.view.Gravity
import android.view.View
import android.widget.*
import androidx.annotation.UiThread
import me.shkschneider.skeleton.R
import me.shkschneider.skeleton.helper.Metrics

// <https://github.com/javiersantos/BottomDialogs>
// app:layout_behavior="android.support.design.widget.BottomSheetBehavior"
class BottomSheet(val builder: Builder) {

    private var dialog: Dialog

    init {
        dialog = Dialog(builder.activity, R.style.SkeletonTheme_BottomSheet)
        val view = builder.activity.layoutInflater.inflate(R.layout.sk_bottomsheet, null)
        val icon = view.findViewById<ImageView>(R.id.sk_bottomsheet_icon)
        icon.visibility = View.VISIBLE
        icon.setImageDrawable(builder.icon)
        val title = view.findViewById<TextView>(R.id.sk_bottomsheet_title)
        title.text = builder.title
        val content = view.findViewById<TextView>(R.id.sk_bottomsheet_content)
        content.text = builder.content
        val customView = view.findViewById<FrameLayout>(R.id.sk_bottomsheet_customView)
        builder.customView?.let {
            customView.addView(it)
            customView.setPadding(builder.paddingLeft, builder.paddingTop, builder.paddingRight, builder.paddingBottom)
        }
        val negative = view.findViewById<Button>(R.id.sk_bottomsheet_cancel)
        builder.negative?.let { text ->
            negative.visibility = View.VISIBLE
            negative.text = text
            negative.setOnClickListener {
                builder.negativeCallback?.onClick(dialog, DialogInterface.BUTTON_NEGATIVE)
                dialog.dismiss()
            }
        }
        val positive = view.findViewById<Button>(R.id.sk_bottomsheet_ok)
        builder.positive?.let { text ->
            positive.visibility = View.VISIBLE
            positive.text = text
            positive.setOnClickListener {
                builder.positiveCallback?.onClick(dialog, DialogInterface.BUTTON_POSITIVE)
                dialog.dismiss()
            }
        }
        dialog.setContentView(view)
        dialog.setCancelable(builder.cancelable)
        dialog.window?.setLayout(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)
        dialog.window?.setGravity(Gravity.BOTTOM)
    }

    @UiThread
    fun show() {
        dialog.show()
    }

    @UiThread
    fun dismiss() {
        dialog.dismiss()
    }

    class Builder(val activity: Activity) {

        var icon: Drawable? = null
        var title: String? = null
        var content: String? = null
        var negative: String? = null
        var positive: String? = null
        var negativeCallback: DialogInterface.OnClickListener? = null
        var positiveCallback: DialogInterface.OnClickListener? = null
        var customView: View? = null
        var paddingLeft = 0
        var paddingTop = 0
        var paddingRight = 0
        var paddingBottom = 0
        var cancelable = false

        init {
            cancelable = true
        }

        fun setTitle(title: String): Builder {
            this.title = title
            return this
        }

        fun setContent(content: String): Builder {
            this.content = content
            return this
        }

        fun setIcon(icon: Drawable): Builder {
            this.icon = icon
            return this
        }

        fun setPositive(positive: String, onClickListener: DialogInterface.OnClickListener?): Builder {
            this.positive = positive
            positiveCallback = onClickListener
            return this
        }

        fun setNegative(negative: String, onClickListener: DialogInterface.OnClickListener?): Builder {
            this.negative = negative
            negativeCallback = onClickListener
            return this
        }

        fun setCancelable(cancelable: Boolean): Builder {
            this.cancelable = cancelable
            return this
        }

        fun setCustomView(customView: View): Builder {
            this.customView = customView
            paddingLeft = 0
            paddingTop = 0
            paddingRight = 0
            paddingBottom = 0
            return this
        }

        fun setCustomView(customView: View, left: Int, top: Int, right: Int, bottom: Int): Builder {
            this.customView = customView
            paddingLeft = Metrics.pixelsFromDp(left.toFloat())
            paddingTop = Metrics.pixelsFromDp(top.toFloat())
            paddingRight = Metrics.pixelsFromDp(right.toFloat())
            paddingBottom = Metrics.pixelsFromDp(bottom.toFloat())
            return this
        }

        @UiThread
        fun build(): BottomSheet {
            return BottomSheet(this)
        }

    }

}
